### ISSUE TITLE
In clustered mode，if do not use acquireTriggersWithinLock properties will appear ABA problem. 

### ISSUE THREAD

**Issue by wenniuwuren** [quartz-scheduler/quartz#107]
In clustered mode，if do not use acquireTriggersWithinLock properties will appear ABA problem. 

here is the quartz source code:
```
public List<OperableTrigger> acquireNextTriggers(final long noLaterThan, final int maxCount, final long timeWindow)
        throws JobPersistenceException {
if(isAcquireTriggersWithinLock() || maxCount > 1) { 
            lockName = LOCK_TRIGGER_ACCESS;
        } else {
            lockName = null;
        }
...
```
```
protected List<OperableTrigger> acquireNextTrigger(Connection conn, long noLaterThan, int maxCount, long timeWindow)
        throws JobPersistenceException {
 
        ... 

        do {
            currentLoopCount ++;
            try {
                List<TriggerKey> keys = getDelegate().selectTriggerToAcquire(conn, noLaterThan + timeWindow, getMisfireTime(), maxCount);
                
                // No trigger is ready to fire yet.
                if (keys == null || keys.size() == 0)
                    return acquiredTriggers;

                long batchEnd = noLaterThan;

                for(TriggerKey triggerKey: keys) {
                    
                    ...

                    // We now have a acquired trigger, let's add to return list.
                    // If our trigger was no longer in the expected state, try a new one.
                    int rowsUpdated = getDelegate().updateTriggerStateFromOtherState(conn, triggerKey, STATE_ACQUIRED, STATE_WAITING);
                    if (rowsUpdated <= 0) {
                        continue; // next trigger
                    }
                    nextTrigger.setFireInstanceId(getFiredTriggerRecordId());
                    getDelegate().insertFiredTrigger(conn, nextTrigger, STATE_ACQUIRED, null);

                    if(acquiredTriggers.isEmpty()) {
                        batchEnd = Math.max(nextTrigger.getNextFireTime().getTime(), System.currentTimeMillis()) + timeWindow;
                    }
                    acquiredTriggers.add(nextTrigger);
                }

                ...

            } catch (Exception e) {
                throw new JobPersistenceException(
                          "Couldn't acquire next trigger: " + e.getMessage(), e);
            }
        } while (true);
        
        // Return the acquired trigger list
        return acquiredTriggers;
    }
```

 In my configuration, i do not set the batchTriggerAcquisitionMaxCount property, so the default value is 1 and maxCount always equals 1. so acquireNextTriggers() method just have the TRIGGERS table row locks. Then the ABA problem will appear, cluster Server_A get trigger_a, cluster Server_B also can get trigger_a in the acquireNextTrigger() method with selectTriggerToAcquire() method, Server_A change trigger_a status WAITING->ACQUIRED->EXECUTING->WAITING, in this situation Server_B execute updateTriggerStateFromOtherState() method will success, finally the trigger_a will be fired twice. 

I think this is important infomation, but clustering depoly document not mention it. Theoretically, set acquireTriggersWithinLock properties is not enough. TRIGGERS table may be need add version_number field to avoid this problem. 

Here is my two server debug log,
Server_A: 

> 2017-02-17 20:58:12.974 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:58:39.996 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:59:07.272 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:59:31.635 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 1 triggers
> 2017-02-17 21:00:00.002 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:107] Lock 'TRIGGER_ACCESS' is desired by: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.002 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:92] Lock 'TRIGGER_ACCESS' is being obtained: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.003 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:116] Lock 'TRIGGER_ACCESS' given to: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.008 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:141] Lock 'TRIGGER_ACCESS' returned by: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.008 [DEBUG] [org.quartz.core.JobRunShell:201] Calling execute on job DEFAULT.jobDetail
> 2017-02-17 21:00:00.013 [DEBUG] [org.apache.http.client.protocol.RequestAddCookies:122] CookieSpec selected: best-match
> 2017-02-17 21:00:00.013 [DEBUG] [org.apache.http.client.protocol.RequestAuthCache:75] Auth cache not set in the context
> 2017-02-17 21:00:00.013 [DEBUG] [org.apache.http.impl.conn.PoolingHttpClientConnectionManager:215] Connection request: [route: {}->http://ip:80][total kept alive: 0; route allocated: 0 of 2; total allocated: 0 of 20]

Server_B:

> 2017-02-17 20:58:11.974 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:58:38.915 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:59:08.128 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 20:59:33.916 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 1 triggers
> 2017-02-17 21:00:00.001 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:107] Lock 'TRIGGER_ACCESS' is desired by: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.001 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:92] Lock 'TRIGGER_ACCESS' is being obtained: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.002 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:116] Lock 'TRIGGER_ACCESS' given to: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.008 [DEBUG] [org.quartz.impl.jdbcjobstore.StdRowLockSemaphore:141] Lock 'TRIGGER_ACCESS' returned by: scheduler_QuartzSchedulerThread
> 2017-02-17 21:00:00.009 [DEBUG] [org.quartz.core.JobRunShell:201] Calling execute on job DEFAULT.jobDetail
> 2017-02-17 21:00:00.013 [DEBUG] [org.quartz.core.QuartzSchedulerThread:276] batch acquisition of 0 triggers
> 2017-02-17 21:00:00.014 [DEBUG] [org.apache.http.client.protocol.RequestAddCookies:122] CookieSpec selected: best-match
> 2017-02-17 21:00:00.014 [DEBUG] [org.apache.http.client.protocol.RequestAuthCache:75] Auth cache not set in the context
> 2017-02-17 21:00:00.014 [DEBUG] [org.apache.http.impl.conn.PoolingHttpClientConnectionManager:215] Connection request: [route: {}->http://ip:80][total kept alive: 0; route allocated: 0 of 2; total allocated: 0 of 20]

My Cron Expression: 00 00 21 17 02 ? 2017

My property configuration:
```
<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://www.springframework.org/schema/beans"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-3.2.xsd">

    <bean id="jobDetail" class="org.springframework.scheduling.quartz.JobDetailFactoryBean">
        <property name="jobClass" value="com.netease.mail.yanxuan.scheduler.task.JobInvokeService" />
        <property name="durability" value="true"/>
    </bean>

    <bean id="scheduler" lazy-init="false" autowire="no"
          class="org.springframework.scheduling.quartz.SchedulerFactoryBean">

        <property name="jobDetails">
            <list>
                <ref bean="jobDetail"/>
            </list>
        </property>

        <property name="dataSource" ref="dataSource" />

        <property name="overwriteExistingJobs" value="true"/>


        <property name="quartzProperties">
            <props>
                <prop key="org.quartz.scheduler.instanceName">EventScheduler</prop>
                <prop key="org.quartz.scheduler.instanceId">AUTO</prop>

                <!-- Configure ThreadPool -->
                <prop key="org.quartz.threadPool.class">org.quartz.simpl.SimpleThreadPool</prop>
                <prop key="org.quartz.threadPool.threadCount">50</prop>
                <prop key="org.quartz.threadPool.threadPriority">5</prop>
                <prop key="org.quartz.threadPool.threadsInheritContextClassLoaderOfInitializingThread">true</prop>


                <!-- Configure JobStore -->
                <prop key="org.quartz.jobStore.class">org.quartz.impl.jdbcjobstore.JobStoreCMT</prop>
                <prop key="org.quartz.jobStore.misfireThreshold">60000</prop>
                <prop key="org.quartz.jobStore.driverDelegateClass">org.quartz.impl.jdbcjobstore.StdJDBCDelegate</prop>
                <prop key="org.quartz.jobStore.tablePrefix">QRTZ_</prop>
                <prop key="org.quartz.jobStore.maxMisfiresToHandleAtATime">10</prop>
                <prop key="org.quartz.jobStore.isClustered">true</prop>
                <prop key="org.quartz.jobStore.clusterCheckinInterval">20000</prop>
                <prop key="org.quartz.jobStore.dontSetAutoCommitFalse">true</prop>
                <prop key="org.quartz.jobStore.txIsolationLevelSerializable">false</prop>
                <prop key="org.quartz.jobStore.useProperties">false</prop>

            </props>
        </property>
        <property name="applicationContextSchedulerContextKey" value="applicationContext" />
    </bean>

</beans>
```
```
<dependency>
        <groupId>org.quartz-scheduler</groupId>
        <artifactId>quartz</artifactId>
        <version>2.2.1</version>
    </dependency>
```

**Comment by zhegexiaohuozi** [comment]
I also met this problem.

**Comment by giilby** [comment]
We also encountered this problem running v2.2.3. From our initial research, it seems that setting `acquireTriggersWithinLock=true` for clustered environments should reliably prevent the race condition. @wenniuwuren, could you explain why you think "Theoretically, set `acquireTriggersWithinLock` properties is not enough"?

**Comment by zhegexiaohuozi** [comment]
@giilby The official documentation say `This was once necessary (in previous versions of Quartz) to avoid dead-locks with particular databases, but is no longer considered necessary, hence the default value is “false”.` about `acquireTriggersWithinLock`([official doc](http://www.quartz-scheduler.org/documentation/quartz-2.x/configuration/ConfigJobStoreTX.html)).Are you sure it is necessary in version 2.2.3 or higher version?

**Comment by wenniuwuren** [comment]
@giilby I mension in this issue that the ABA problem will make set `acquireTriggersWithinLock=true` property not enough to guarantee two or more server fired trigger just one time.

**Comment by jhouserizer** [comment]
For those encountering a problem, can you clarify the transaction isolation level that you're using/getting from your database?

**Comment by giilby** [comment]
@zhegexiaohuozi Without `acquireTriggersWithinLock=true`, I was able to locally replicate an issue where two cluster nodes would occasionally run the same job at the same time. Per
[this discussion](http://quartz.10975.n7.nabble.com/Locking-around-acquireNextTrigger-necessary-td1449.html), it seems that locking around trigger acquisition was considered unnecessary due to an optimistic locking strategy when setting `trigger_state` to `ACQUIRED`. However, it seems that there are rare cases when, as described by @wenniuwuren, two nodes can select the same `WAITING` trigger, then one node will commit all status changes ( `WAITING -> ACQUIRED -> EXECUTING -> WAITING` before the other even gets to the optimistic lock. I will post more details below.

**Comment by giilby** [comment]
@wenniuwuren Right, but why do you think that? From my investigation and testing, `acquireTriggersWithinLock=true` should fix the ABA issue because only one node will actually obtain the `WAITING` trigger via a `SELECT`.

**Comment by giilby** [comment]
@jhouserizer We're using `READ COMMITTED`. I've attached [some logs from MySQL](https://github.com/quartz-scheduler/quartz/files/901215/mysql-general-log.txt)
 (taken with `general_log = 'ON'`) that I filtered/edited to narrow in on the problem. Let me know if you need more details about my test setup, but these logs illustrate how `test-job-6` executed on both cluster nodes. Maybe it's possible to fix this by tweaking the optimistic lock (`UPDATE qrtz_triggers SET trigger_state = 'ACQUIRED' WHERE ... AND trigger_state = 'WAITING'`) to consider the expected `next_fire_time` as well?

**Comment by wenniuwuren** [comment]
@jhouserizer We are using `REPEATABLE-READ`.

**Comment by zhegexiaohuozi** [comment]
@jhouserizer ours
```
mysql> select @@global.tx_isolation;
+-----------------------+
| @@global.tx_isolation |
+-----------------------+
| REPEATABLE-READ       |
+-----------------------+
1 row in set
```

**Comment by gdpotter** [comment]
I _think_ I'm having the same issue mentioned above (although I am using postgres). I posted my details in the quartz forums https://groups.google.com/forum/#!category-topic/quartz-scheduler/jdbcjobstore/2sHeaufVIl4

**Comment by ihorpts** [comment]
I guess this problem can be reproducible with https://github.com/quartz-scheduler/quartz/issues/29#issuecomment-378911357 app. Putting `acquireTriggersWithinLock=true` + `@DisallowConcurrentExecution` makes the problem vanished in deterministic way for me.

**Comment by Dorae132** [comment]
@这里问题的关键应该有以下两步：
1、获取waitting状态的trigger，修改trigger状态为acquired，同时生成firedTrigger，然后修改trigger状态为waiting/blocked
2、触发firedtrigger，根据是否为最后一次触发决定是否修改trigger状态为complete。
所以避免并行执行的关键在于，对于同一个trigger，firedTrigger不能同时存在。对于1，2两步获取锁的问题，其都为lock_trigger_access锁，但是在其执行完后都会释放锁。
所以对于获取trigger时加锁（第1步）是不能解决问题的，因为最终只有更新其状态为acquired成功的线程获得继续的权力，这里加锁只能在一定程度上防并发；
但是对于2已经执行完的任务，如果为statelessJob, 那么这里是存在竞争条件的，也就是在这里，可以继续生成新的firedTrigger，从而导致的并行执行（也就是ABA问题），但是@DisAllowConcurrent注解（statefuleJob）应该可以解决该问题。 @giilby @wenniuwuren

**Comment by Dorae132** [comment]
There are two key steps here in my investigation.
  First to get the triggers in waiting state, modify the triggers state to be acquired, generate the firedTriggers, and the modify the triggers state to waiting or blocked.
  Then the firedTriggers are fired, and modify the triggers to completion status if it is the last trigger.
  If there is only one firedTrigger for the same trigger at the same time, the issue will never reproduce.
  So it seems that locking around trigger acquisition was considered unnecessary due to an optimistic locking strategy when setting trigger_state to ACQUIRED. But the ABA will occur, because there will be two or more firedTriggers for the same trigger.

**Comment by slankka** [comment]
The ```acquireTriggersWithinLock=true``` works, need not ```@DisallowConcurrentExecution```.

**Comment by wangfan9002** [comment]
I think  can be done check the trigger's state and nextFireTime as the same time. 


// Make sure trigger wasn't deleted, paused, or completed...
        try { // if trigger was deleted, state will be STATE_DELETED
            String state = getDelegate().selectTriggerState(conn,
                    trigger.getKey());
            if (!state.equals(STATE_ACQUIRED)) {
                return null;
            }
        } catch (SQLException e) {
            throw new JobPersistenceException("Couldn't select trigger state: "
                    + e.getMessage(), e);
        }

**Comment by zemian** [comment]
Hi all, sorry about the long delay on responding on this ticket.

I think the this issue is more DB specific, and it can be resolve by `acquireTriggersWithinLock=true`; as many of you already discovered. The flag is there to ensure if DB that does exhibits this behavior, you can use it guarantee to avoid the problem. The draw back is performance obviously.

It would be interesting to collect and see which DB specific with their isolation level that shows this exact problems.

**Comment by zemian** [comment]
Hi all, I did some searching into this issue. It looks like with Postgres, the isolation level READ_COMMITED (which is default) should run without problem when acquire for triggers. You can read more here: https://www.postgresql.org/docs/9.1/transaction-iso.html (I believe @gdpotter mentioned above, he is using Postgres, but his problem is slightly different, and which is also resolved by another ticket #189 already)

For other DB, I encourage you to read up on their default isolation level and their implementation behavior on how it handle concurrent update behavior. For example the sql to acquire triggers is done like this: 
````
UPDATE QRTZ_TRIGGERS SET TRIGGER_STATE = 'ACQUIRED' WHERE SCHED_NAME = 'MYSCHEDULER'
 AND TRIGGER_NAME = 'T1' AND 
TRIGGER_GROUP = 'DEFAULT' AND TRIGGER_STATE = 'WAITING'
````
To be safe, you can always use the `acquireTriggersWithinLock=true` flag, which will incur some performance, but should resolve this issue.

**Comment by way2nikhil** [comment]
Commenting on closed thread , but wanted to understand if this reproduces only when recoveryrequested= true or otherwise also.
@wenniuwuren @giilby @jhouserizer 
btw we are using oracle database seeing this issue when recoveryrequested= true and acquireTriggersWithinLock=true and vm time are in sync.

**Comment by fan0916** [comment]
we get this issue too,we use xxl-job(2.0.2) which relies on quartz 2.3.1
according to the server debug log posted by @wenniuwuren, it's very hard to say it's ABA problem.
after retriving triggers, thread sleep 20 seconds, and then goto method triggersFired, the log of acquiring lock TRIGGER_ACCESS indicates this

so in one process for a trigger to finish waiting->acquired->waiting, it takes at least 20 seconds. which is abviously longer than the excuting time of another process's acquireNextTriggers

according above infomation, it seems very likely to be database isolation level issue, but i dont't think so

i have tested our database(mysql read-commited), compareAndSwap concurrent update didn't make mistake

let's turn back to ABA direction, what if the possible issue is : 
one process acquireNextTriggers and update trigger stage to acquired, and in the later 20s, another process updated trigger stage from acquired to waiting by cas or just force update(i can't locate this kind of code, just guess). then later process goes into acquireNextTriggers and successfully acquire trigger.if this is the update path of replicate trigger, add lock may not help. but according to @slankka post, this config works, there maybe another story

would be appreciated if anyone can help and verify this(figure out the guess update)

**Comment by fan0916** [comment]
our issue prove to be aba problem, one process finishes wating->acquired->waiting in the gap of another process wating -> acquired

here is the sql log:


SQL Statement | Thread ID | Operation | Status | Time   Consumption(ms) | Executed At | Execute Time (ms) | Updated Rows | Scanned Rows
-- | -- | -- | -- | -- | -- | -- | -- | --
UPDATE XXL_JOB_QRTZ_TRIGGERS SET JOB_NAME = '35',   JOB_GROUP = 'DEFAULT', DESCRIPTION = null, NEXT_FIRE_TIME = 1587952800000,   PREV_FIRE_TIME = 1587952200000, TRIGGER_STATE = 'WAITING', TRIGGER_TYPE =   'CRON', START_TIME = 1585288184000, END_TIME = 0, CALENDAR_NAME = null,   MISFIRE_INSTR = 2, PRIORITY = 5 WHERE SCHED_NAME = 'getSchedulerFactoryBean'   AND TRIGGER_NAME = '35' AND TRIGGER_GROUP = 'DEFAULT' | 11296710 | UPDATE | Completed | 0.17 | 2020/4/27 9:50 | 2020-04-27 09:50:00.150 | 1 | 1
UPDATE   XXL_JOB_QRTZ_TRIGGERS SET TRIGGER_STATE = 'ACQUIRED' WHERE SCHED_NAME =   'getSchedulerFactoryBean' AND TRIGGER_NAME = '35' AND TRIGGER_GROUP =   'DEFAULT' AND TRIGGER_STATE = 'WAITING' | 11296710 | UPDATE | Completed | 0.19 | 2020/4/27 9:50 | 2020-04-27 09:50:00.141 | 1 | 1
UPDATE   XXL_JOB_QRTZ_TRIGGERS SET TRIGGER_STATE = 'ACQUIRED' WHERE SCHED_NAME =   'getSchedulerFactoryBean' AND TRIGGER_NAME = '35' AND TRIGGER_GROUP =   'DEFAULT' AND TRIGGER_STATE = 'WAITING' | 12177889 | UPDATE | Completed | 0.39 | 2020/4/27 9:50 | 2020-04-27 09:50:00.277 | 1 | 1
UPDATE   XXL_JOB_QRTZ_TRIGGERS SET JOB_NAME = '35', JOB_GROUP = 'DEFAULT', DESCRIPTION   = null, NEXT_FIRE_TIME = 1587952800000, PREV_FIRE_TIME = 1587952200000,   TRIGGER_STATE = 'WAITING', TRIGGER_TYPE = 'CRON', START_TIME = 1585288184000,   END_TIME = 0, CALENDAR_NAME = null, MISFIRE_INSTR = 2, PRIORITY = 5 WHERE   SCHED_NAME = 'getSchedulerFactoryBean' AND TRIGGER_NAME = '35' AND   TRIGGER_GROUP = 'DEFAULT' | 12177889 | UPDATE | Completed | 0.39 | 2020/4/27 9:50 | 2020-04-27 09:50:00.369 | 1 | 1
